app/lti.go

```diff
@@ -4,14 +4,48 @@
 package app
 
 import (
+	"encoding/json"
+	"fmt"
 	"net/http"
 
+	"github.com/pkg/errors"
+
 	"github.com/mattermost/mattermost-server/v5/model"
 	"github.com/mattermost/mattermost-server/v5/shared/mlog"
 )
 
+//
+//	GetLTISettings() reads the LTI Config from Plugin Config
+//
+func (a *App) GetLTISettings() (*model.LTISettings, error) {
+
+	//	Check that if a plugin with LTI_PLUGIN_ID is installed in the server.
+	var LTIConfig map[string]interface{}
+	LTIConfig, ok := a.Config().PluginSettings.Plugins[model.LTI_PLUGIN_ID]
+	if !ok {
+		return nil, fmt.Errorf("LTI Configuration Plugin not found")
+	}
+
+	configJson, err := json.Marshal(LTIConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error marshaling LTI Config: %s")
+	}
+
+	var LTISettings *model.LTISettings
+	if err = json.Unmarshal(configJson, &LTISettings); err != nil {
+		return nil, errors.Wrap(err, "Error unmarshaling LTI Config from json: %s")
+	}
+	return LTISettings, nil
+}
```

Comment:
From what I can figure out the reason for this Marshal and then UnMarshal of the config is specifically because you have changed the properties names of the LTISettings, usually just lowercasing them. I think I'd prefer leaving the names as they were cased before.

From what I can tell there is no requirement that the plugin property names be lowercase and not mixed case.

Reply:
Actually, the reason for `Marshal` and `Unmarshal` is because the type of `PluginSettings.Plugins` is `map[string]interface{}`. and here we want to convert `map[string]interface{}` to `LTISettings` Struct.

Reply:
Also, I don't think it is possible for plugin property to be mixed-case, they are lowercase always.

Reply:
I looked through the mattermost code that read the plugin config and I didn't find anything that would require that plugin properties be mixed case. If you could point me at the mattermost code that requires lowercase, I'd appreciate it.

Reply:
so in the mattermost webapp code, we change the plugin property to lowercase [here](https://github.com/mattermost/mattermost-webapp/blob/4486ca91fe5dec6e6f34fba01a9c1790beeab335/components/admin_console/custom_plugin_settings/index.js#L38). 
But even if we make the plugin property mixed-case, we would still need to do the `Marshal` and `Unmarshal` step to
convert `map[string]interface{}` to `LTISettings` Struct.

Reply:
> On Tue, Jun 1, 2021 at 8:01 AM @mkdbns wrote:
> Hi Mike,
> I saw the thread and comments on the LTI config JSON keys being lowercase. I am not able to leave comments there so sending this via email. 
> The lowercase properties are used for all MM plugin properties.  It is actually done in the MM code here:
> 
> https://github.com/mattermost/mattermost-webapp/blob/4486ca91fe5dec6e6f34fba01a9c1790beeab335/components/admin_console/custom_plugin_settings/index.js#L38
> I would not recommend trying to force mixed-case just for the LTI settings. This would likely take product code changes and may cause other side effects now or in the future.
> Does that make sense?
> All the best,
> Mike

Hi @mkdbns,Thanks, that pointer to where in the code mattermost "required" the plugin properties to be lowercase is exactly what I wanted to know. (Also you should now have the appropriate permissions to add comments to the PRs)

I had done some googling and found Mattermost [plugin documentation](https://developers.mattermost.com/extend/plugins/best-practices/) but none of it mentioned that plugin properties must be lowercase or would be converted to lowercase to be saved in the config file.

I found the [plugin API](https://developers.mattermost.com/extend/plugins/server/reference/) for accessing the config  using `GetConfig` and `SaveConfig`, and I looked at the Mattermost code for that, and nowhere did I see anything that implied the properties needed to be lowercase.

However I did not look at _mattermost-webapp_.

So if it was *only* because of convention that we were using lowercase for the lti plugin config, then I felt it was worth breaking the convention in this case.

Now that I see where Mattermost is enforcing this lowercase property names (I'm not sure why it does, but that isn't important here), I agree that there is no need for us to attempt to work around this restriction.

That said @hrishavkmr I am almost positive that it is not required to use Marshal/UnMarshal to assign an interface{}  to a struct (`LTISettings`) in the go language. I believe that either a [type assertion or a type conversion](https://www.sohamkamani.com/golang/type-assertions-vs-type-conversions/) should do it, I just haven't had a chance to test to figure out which should be used and the official documentation I've read hasn't be totally clear. This is meant only to address your previous comment, since we do need to have the properties be lowercase in the config, Marshal/UnMarshal seems the proper way to maintain the correct name mapping.

I will continue my review with this in mind.